app/services/transaction_service/transaction.rb

```diff
@@ -100,30 +102,48 @@ def create(opts, force_sync: true)
   end
 
   def finalize_create(community_id:, transaction_id:, force_sync: true)
-    tx = TxStore.get_in_community(community_id: community_id, transaction_id: transaction_id)
-
-    if tx.nil?
-
-      # Transaction doesn't exist.
-      #
-      # This may happen if the finalize_create action has been called already, and it failed.
-      # If the finalization fails (e.g. booking fails), we void the payment and delete the
-      # transaction.
-
-      return Result::Error.new("Can't find transaction, id: #{transaction_id}, community_id: #{community_id}", {code: :tx_not_existing})
-    end
-
-    tx_process = tx_process(tx[:payment_process])
-    gw = gateway_adapter(tx[:payment_gateway])
-
-    res = tx_process.finalize_create(
-      tx: tx,
-      gateway_adapter: gw,
-      force_sync: force_sync)
+    # Transaction may be nil, if it has been deleted due to voided payment
+    m_tx = Maybe(TxStore.get_in_community(community_id: community_id, transaction_id: transaction_id))
+
+    proc_response =
+      if !force_sync
+        # Try to find existing process token
+        # This may happen if finalize_create action has been called already, for example
+        # as a reaction to payment event
+
+        proc_token = ProcessTokenStore.get_by_transaction(community_id: community_id,
+                                                          transaction_id: transaction_id,
+                                                          op_name: :do_finalize_create)
+
+        proc_status_response(proc_token) if proc_token
+      end
+
+    res =
+      if proc_response
+        proc_response
+      elsif m_tx.is_none?
+
+        # Transaction doesn't exist.
+        #
+        # This may happen if the finalize_create action has been called already, and it failed.
+        # If the finalization fails (e.g. booking fails), we void the payment and delete the
+        # transaction.
+        Result::Error.new("Can't find transaction, id: #{transaction_id}, community_id: #{community_id}", {code: :tx_not_existing})
+      else
+        tx = m_tx.get
+
+        tx_process = tx_process(tx[:payment_process])
+        gw = gateway_adapter(tx[:payment_gateway])
+
+        tx_process.finalize_create(
+          tx: tx,
+          gateway_adapter: gw,
+          force_sync: force_sync)
+      end
```

Comment:
This logic could be simplified. :) There's nested ifs and m.get / m.is_none? used which smells like there's room for improvement.